# Indexing Loans

Read the loan data in from the CSV file and utilize it to answer various questions about the individuls taking our 3 year loans. 

Instructions:

1. Using the Pandas `read_csv` function and the Path module, read in the `loans.csv` file and create the Pandas DataFrame. Review the first five rows of the resulting DataFrame.

2. Using `iloc[]`, show the first `10` records of the data.

3. Generate the summary statistics for all of the `loans.csv` data.

4. Using `iloc[]`, create a subset DataFrame `filtered_df` by selecting only the following columns:

    * `loan_amnt`
    * `term`
    * `int_rate`
    * `emp_title`
    * `annual_inc`
    * `purpose`

5. Using `loc[]`, filter `filtered_df` by row values where `term` is equal to `36 months` in order to focus on only three-year loan records.

6. Modify rows with `term` values equal to `36 months` to be `3 years`.

7. Use the `isnull` and `sum` function to evaluate the number of missing values in the `term_df` DataFrame. Use the `fillna` function to replace the NaN values with 'Unknown'. Review the first five rows of the new DataFrame

8. Generate the summary statistics for `term_df` after all modifications.

9. Use the `value_counts()` function on the `emp_title` column of the `term_df` DataFrame to see the unique value counts for employee titles of three-year loan customers.

10. Use the `value_counts()` function on the `purpose` column of the `term_df` DataFrame to see the unique value counts for loan purposes of three-year loan customers.

11. Filter `term_df` by rows with `annual_inc` greater than `80000`. Use the `describe` function to see the mean `int_rate` of three-year loan customers with annual incomes greater than $80,000.

12. Filter `term_df` by rows with `annual_inc` less than `80000`. Use the `describe` function to see the average `int_rate` of three-year loan customers with annual incomes less than $80,000.

13. Answer the following questions about individuals taking out 3 year loans:

    * What kind of customers (employee title) seem to ask for three-year loans most frequently?

    * What are three-year loans generally used for?

    * What is the difference between counts of three-year loan customers with annual incomes greater than 80,000, compared to those with annual incomes less than 80,000?

    * What is the difference between interest rates for customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000?

In [1]:
# Import libraries and dependencies
import pandas as pd
from pathlib import Path

## Using the Pandas `read_csv` function and the Path module, read in in the `loans.csv` file and create the Pandas DataFrame. Review the first five rows of the resulting DataFrame.

In [2]:
# Read in the CSV as a DataFrame
loans_df = pd.read_csv(Path('../Resources/loans.csv'))

# Review the first five rows of the DataFrame
loans_df.head()

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  installment  \
0      10000        10000          10000.0  36 months    0.1033       324.23   
1       4000         4000           4000.0  36 months    0.2340       155.68   
2       5000         5000           5000.0  36 months    0.1797       180.69   
3       9600         9600           9600.0  36 months    0.1298       323.37   
4       2500         2500           2500.0  36 months    0.1356        84.92   

  grade sub_grade       emp_title emp_length home_ownership  annual_inc  \
0     B        B1             NaN   < 1 year       MORTGAGE    280000.0   
1     E        E1        Security    3 years           RENT     90000.0   
2     D        D1  Administrative    6 years       MORTGAGE     59280.0   
3     B        B5             NaN        NaN       MORTGAGE     35704.0   
4     C        C1            Chef  10+ years           RENT     55000.0   

  verification_status issue_d loan_status pymnt_plan             purpose  
0        Not Verified  Dec-18     Current          n  debt_consolidation  
1     Source Verified  Dec-18     Current          n  debt_consolidation  
2     Source Verified  Dec-18     Current          n  debt_consolidation  
3        Not Verified  Dec-18     Current          n    home_improvement  
4        Not Verified  Dec-18     Current          n  debt_consolidation

## Using the `iloc[]` function, show the first `10` records of the data.

In [3]:
# Retrieve rows with index 0 up to 10 (not including)
loans_df.iloc[0:10]

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  installment  \
0      10000        10000          10000.0  36 months    0.1033       324.23   
1       4000         4000           4000.0  36 months    0.2340       155.68   
2       5000         5000           5000.0  36 months    0.1797       180.69   
3       9600         9600           9600.0  36 months    0.1298       323.37   
4       2500         2500           2500.0  36 months    0.1356        84.92   
5      30000        30000          30000.0  60 months    0.1894       777.23   
6      23000        23000          23000.0  60 months    0.2089       620.81   
7      32075        32075          32075.0  60 months    0.1180       710.26   
8       8000         8000           8000.0  36 months    0.2340       311.35   
9      10000        10000          10000.0  60 months    0.1992       264.50   

  grade sub_grade           emp_title emp_length home_ownership  annual_inc  \
0     B        B1                 NaN   < 1 year       MORTGAGE    280000.0   
1     E        E1            Security    3 years           RENT     90000.0   
2     D        D1      Administrative    6 years       MORTGAGE     59280.0   
3     B        B5                 NaN        NaN       MORTGAGE     35704.0   
4     C        C1                Chef  10+ years           RENT     55000.0   
5     D        D2         Postmaster   10+ years       MORTGAGE     90000.0   
6     D        D4            Operator    5 years           RENT     68107.0   
7     B        B4  Nursing Supervisor  10+ years       MORTGAGE    150000.0   
8     E        E1             Manager  10+ years            OWN     43000.0   
9     D        D3    Material Handler  10+ years       MORTGAGE     80000.0   

  verification_status issue_d loan_status pymnt_plan             purpose  
0        Not Verified  Dec-18     Current          n  debt_consolidation  
1     Source Verified  Dec-18     Current          n  debt_consolidation  
2     Source Verified  Dec-18     Current          n  debt_consolidation  
3        Not Verified  Dec-18     Current          n    home_improvement  
4        Not Verified  Dec-18     Current          n  debt_consolidation  
5     Source Verified  Dec-18     Current          n  debt_consolidation  
6     Source Verified  Dec-18     Current          n  debt_consolidation  
7        Not Verified  Dec-18     Current          n         credit_card  
8     Source Verified  Dec-18     Current          n  debt_consolidation  
9        Not Verified  Dec-18     Current          n  debt_consolidation

## Generate the summary statistics for all of the `loans_df` DataFrame.

In [4]:
# Generate the summary statistics for the DataFrame
# Be sure to include all the columns
loans_df.describe(include='all')

loan_amnt    funded_amnt  funded_amnt_inv       term  \
count   128412.000000  128412.000000    128412.000000     128412   
unique            NaN            NaN              NaN          2   
top               NaN            NaN              NaN  36 months   
freq              NaN            NaN              NaN      88179   
mean     15971.321021   15971.321021     15968.498166        NaN   
std      10150.384233   10150.384233     10152.168970        NaN   
min       1000.000000    1000.000000       725.000000        NaN   
25%       8000.000000    8000.000000      8000.000000        NaN   
50%      14000.000000   14000.000000     14000.000000        NaN   
75%      21600.000000   21600.000000     21600.000000        NaN   
max      40000.000000   40000.000000     40000.000000        NaN   

             int_rate    installment   grade sub_grade emp_title emp_length  \
count   128412.000000  128412.000000  128412    128412    107465     116708   
unique            NaN            NaN       7        35     43892         11   
top               NaN            NaN       A        A4   Teacher  10+ years   
freq              NaN            NaN   38011      9770      2090      38826   
mean         0.129302     463.259353     NaN       NaN       NaN        NaN   
std          0.050904     285.718256     NaN       NaN       NaN        NaN   
min          0.060000      30.480000     NaN       NaN       NaN        NaN   
25%          0.084600     253.630000     NaN       NaN       NaN        NaN   
50%          0.118000     382.905000     NaN       NaN       NaN        NaN   
75%          0.161400     622.680000     NaN       NaN       NaN        NaN   
max          0.309900    1618.240000     NaN       NaN       NaN        NaN   

       home_ownership    annual_inc verification_status issue_d loan_status  \
count          128412  1.284120e+05              128412  128412      128412   
unique              4           NaN                   3       3           7   
top          MORTGAGE           NaN        Not Verified  Oct-18     Current   
freq            63490           NaN               58350   46305      119540   
mean              NaN  8.279733e+04                 NaN     NaN         NaN   
std               NaN  1.082985e+05                 NaN     NaN         NaN   
min               NaN  0.000000e+00                 NaN     NaN         NaN   
25%               NaN  4.705800e+04                 NaN     NaN         NaN   
50%               NaN  6.800000e+04                 NaN     NaN         NaN   
75%               NaN  9.900000e+04                 NaN     NaN         NaN   
max               NaN  9.757200e+06                 NaN     NaN         NaN   

       pymnt_plan             purpose  
count      128412              128412  
unique          1                  12  
top             n  debt_consolidation  
freq       128412               70603  
mean          NaN                 NaN  
std           NaN                 NaN  
min           NaN                 NaN  
25%           NaN                 NaN  
50%           NaN                 NaN  
75%           NaN                 NaN  
max           NaN                 NaN

## Using `iloc[]`, create a subset DataFrame `filtered_df` by selecting only the following columns:

* `loan_amnt`
* `term`
* `int_rate`
* `emp_title`
* `annual_inc`
* `purpose`

In [5]:
# Using the `iloc` function, create a DataFrame that consists of all rows of the columns:
# loan_amnt, term, int_rate, emp_title, annual_inc and purpose
filtered_df = loans_df.iloc[:, [0, 3, 4, 8, 11, 16]]

# Review the first five rows of the filtered DataFrame
filtered_df.head()

loan_amnt       term  int_rate       emp_title  annual_inc  \
0      10000  36 months    0.1033             NaN    280000.0   
1       4000  36 months    0.2340        Security     90000.0   
2       5000  36 months    0.1797  Administrative     59280.0   
3       9600  36 months    0.1298             NaN     35704.0   
4       2500  36 months    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

## Using `loc[]`, filter `filtered_df` by row values where `term` is equal to `36 months` in order to focus on only three-year loan records.

In [6]:
# Conditional indexing to filter DataFrame where 'term' is equal to '36 months'
term_df = filtered_df.loc[filtered_df['term'] == '36 months'].copy()

# Review the first five rows of the term_df DataFrame
term_df.head()

loan_amnt       term  int_rate       emp_title  annual_inc  \
0      10000  36 months    0.1033             NaN    280000.0   
1       4000  36 months    0.2340        Security     90000.0   
2       5000  36 months    0.1797  Administrative     59280.0   
3       9600  36 months    0.1298             NaN     35704.0   
4       2500  36 months    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

## Modify rows with `term` values equal to `36 months` to be `3 years`.

In [7]:
# Change row values within the 'term' column from '36 months' to '3 Years'
term_df.loc[term_df['term'] == '36 months', 'term'] = '3 Years'

# Review the first five rows of the term_df DataFrame
term_df.head()

loan_amnt     term  int_rate       emp_title  annual_inc  \
0      10000  3 Years    0.1033             NaN    280000.0   
1       4000  3 Years    0.2340        Security     90000.0   
2       5000  3 Years    0.1797  Administrative     59280.0   
3       9600  3 Years    0.1298             NaN     35704.0   
4       2500  3 Years    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

## Use the `isnull` and `sum` function to evaluate the number of missing values in the `term_df` DataFrame. Use the `fillna` function to replace the NaN values with 'Unknown'. Review the first five rows of the new DataFrame 

In [8]:
# Use the isnaull and sum functions to evaluate the number of missing values in the `term_df` DataFrame.
term_df.isnull().sum()

loan_amnt         0
term              0
int_rate          0
emp_title     15669
annual_inc        0
purpose           0
dtype: int64

In [9]:
#  Use the `fillna` function to replace the NaN values with 'Unknown'
term_df = term_df.fillna('Unknown')

# Review the first five rows of the cleaned term_df DataFrame
term_df.head()

loan_amnt     term  int_rate       emp_title  annual_inc  \
0      10000  3 Years    0.1033         Unknown    280000.0   
1       4000  3 Years    0.2340        Security     90000.0   
2       5000  3 Years    0.1797  Administrative     59280.0   
3       9600  3 Years    0.1298         Unknown     35704.0   
4       2500  3 Years    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

### Generate the summary statistics for `term_df` after all modifications.

In [10]:
# Describe summary statistics for three-year loans
term_df.describe()

loan_amnt      int_rate    annual_inc
count  88179.000000  88179.000000  8.817900e+04
mean   13279.580456      0.120610  8.043616e+04
std     9479.610905      0.048955  1.109251e+05
min     1000.000000      0.060000  0.000000e+00
25%     6000.000000      0.078400  4.500000e+04
50%    10000.000000      0.110600  6.500000e+04
75%    18000.000000      0.150200  9.500000e+04
max    40000.000000      0.309900  9.000000e+06

## Use the `value_counts()` function on the `emp_title` column of the `term_df` DataFrame to see the unique value counts for employee titles of three-year loan customers.

In [11]:
# Calculate unique values and counts for employee titles of 3 year customer loans
term_df['emp_title'].value_counts()

Unknown                                   15669
Teacher                                    1433
Manager                                    1207
Registered Nurse                            672
Driver                                      654
                                          ...  
Delivery driver/installer                     1
Human Capital Specialist                      1
City carier                                   1
QA lead                                       1
Sr. Public Health Research Coordinator        1
Name: emp_title, Length: 31470, dtype: int64

## Use the `value_counts()` function on the `purpose` column of the `term_df` DataFrame to see the unique value counts for loan purposes of three-year loan customers.

In [12]:
# Calculate unique values and counts for loan purposes of 3 year customer loans
term_df['purpose'].value_counts()

debt_consolidation    45456
credit_card           24780
other                  5836
home_improvement       5446
major_purchase         1795
medical                1245
car                     838
small_business          828
vacation                749
house                   596
moving                  547
renewable_energy         63
Name: purpose, dtype: int64

## Filter `term_df` by rows with `annual_inc` greater than `80000`. Use the `describe` function to see the mean `int_rate` of three-year loan customers with annual incomes greater than $80,000.

In [13]:
# Display summary statistics where annual income is greater than $80,000 to find count and mean
term_df.loc[term_df['annual_inc'] > 80000].describe()

loan_amnt      int_rate    annual_inc
count  29748.000000  29748.000000  2.974800e+04
mean   18240.116142      0.114116  1.389634e+05
std    10991.909117      0.048067  1.752361e+05
min     1000.000000      0.060000  8.000100e+04
25%    10000.000000      0.075600  9.500000e+04
50%    15700.000000      0.104700  1.120000e+05
75%    25000.000000      0.144700  1.500000e+05
max    40000.000000      0.309400  9.000000e+06

## Filter `term_df` by rows with `annual_inc` less than `80000`. Use the `describe` function to see the average `int_rate` of three-year loan customers with annual incomes less than $80,000.

In [14]:
# Display summary statistics where annual income is less than $80,000 to find count and mean
term_df.loc[term_df['annual_inc'] < 80000].describe()

loan_amnt      int_rate    annual_inc
count  56015.000000  56015.000000  56015.000000
mean   10607.487727      0.124268  49372.805964
std     7318.791900      0.049137  16669.098821
min     1000.000000      0.060000      0.000000
25%     5000.000000      0.081900  37000.000000
50%     9600.000000      0.113100  50000.000000
75%    14000.000000      0.150200  62635.500000
max    40000.000000      0.309900  79999.000000

## Answer the following questions about individuals taking out 3 year loans:

1. What kind of customers (employee title) seem to ask for three-year loans most frequently?
2. What are three-year loans generally used for?
3. What is the difference in count of three-year loan customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000?
4. What is the difference in interest rates of three-year loan customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000? 

1. Interestingly enough Teachers and Managers seem to be the top two contenders for requesting three-year loans. In addition, there are a disproportionate amount of 'Unknown' or un-reported employee titles. Unless this field is optional, this could signal a potential data quality problem.

2. From the value counts of loan purposes, it looks like three-year loans are generally used to consolidate debt.

3. A majority of three-year loan customers have annual incomes of less than 80,000. Counts of customers with annual incomes less than 80,000 compared to greater than 80,000 stand at 56,015 customers vs 29,748 customers, respectively.

4. Interest rates between customers with annual incomes greater than 80,000 compared to those with less than 80,000 differ by approximately 1%; Interest rates were 11.41% and 12.43%, respectively.